dali/core/mm/async_pool_test.cu

```diff
@@ -0,0 +1,325 @@
+
+namespace dali {
+namespace mm {
+
+struct GPUHog {
+  ~GPUHog() {
+    if (mem) {
+      CUDA_DTOR_CALL(cudaFree(mem));
+      mem = nullptr;
+    }
+  }
+
+  void init() {
+    if (!mem)
+      CUDA_CALL(cudaMalloc(&mem, size));
+  }
+
+  void run(cudaStream_t stream, int count = 1) {
+    for (int i = 0; i < count; i++) {
+      CUDA_CALL(cudaMemsetAsync(mem, i+1, size, stream));
+    }
+  }
+
+  uint8_t *mem = nullptr;
+  size_t size = 16<<20;
+};
+
+TEST(MMAsyncPool, SingleStreamReuse) {
+  GPUHog hog;
+  hog.init();
+
+  CUDAStream stream = CUDAStream::Create(true);
+  test::test_device_resource upstream;
+
+  async_pool_base<memory_kind::device, free_tree, std::mutex> pool(&upstream);
+  stream_view sv(stream);
+  int size1 = 1<<20;
+  void *ptr = pool.allocate_async(size1, sv);
+  hog.run(stream, 2);
+  pool.deallocate_async(ptr, size1, sv);
+  void *p2 = pool.allocate_async(size1, sv);
+  CUDA_CALL(cudaStreamSynchronize(stream));
+  EXPECT_EQ(ptr, p2);
+}
+
+__global__ void Check(const void *ptr, size_t size, uint8_t fill, int *failures) {
+  size_t idx = static_cast<size_t>(blockIdx.x) * blockDim.x + threadIdx.x;
+  if (idx < size) {
+    if (static_cast<const uint8_t*>(ptr)[idx] != fill)
+      atomicAdd(failures, 1);
+  }
+}
+
+namespace {
+
+struct block {
+  void *ptr;
+  size_t size;
+  uint8_t fill;
+  cudaStream_t stream;
+};
+
+template <typename Pool, typename Mutex>
+void AsyncPoolTest(Pool &pool, vector<block> &blocks, Mutex &mtx, CUDAStream &stream,
+                   int max_iters = 100000, bool use_hog = false) {
+  stream_view sv(stream);
+  std::mt19937_64 rng(12345);
+  std::poisson_distribution<> size_dist(1024);
+  const int max_size = 1<<20;
+  std::uniform_int_distribution<> sync_dist(10, 10);
+  std::bernoulli_distribution action_dist;
+  std::bernoulli_distribution hog_dist(0.05f);
+  std::uniform_int_distribution<> fill_dist(1, 255);
+  DeviceBuffer<int> failure_buf;
+  int failures = 0;
+  failure_buf.from_host(&failures, 1, stream);
+  GPUHog hog;
+  if (use_hog)
+    hog.init();
+  int hogs = 0;
+  int max_hogs = sync_dist(rng);
+  CUDAEvent event = CUDAEvent::Create();
+  for (int i = 0; i < max_iters; i++) {
+    if (use_hog && hog_dist(rng)) {
+      if (hogs++ > max_hogs) {
+        CUDA_CALL(cudaStreamSynchronize(stream));
+        max_hogs = sync_dist(rng);
+      }
+      hog.run(stream);
+    }
+    if (action_dist(rng) || blocks.empty()) {
+      size_t size;
+      do {
+        size = size_dist(rng);
+      } while (size > max_size);
+      uint8_t fill = fill_dist(rng);
+      void *ptr = stream ? pool.allocate_async(size, sv) : pool.allocate(size);
+      CUDA_CALL(cudaMemsetAsync(ptr, fill, size, stream));
+      {
+        std::lock_guard<Mutex> guard(mtx);
+        blocks.push_back({ ptr, size, fill, stream });
+      }
+    } else {
+      block blk;
+      {
+        std::lock_guard<Mutex> guard(mtx);
+        if (blocks.empty())
+          continue;
```

Comment:
According to L114 it can't happen, right?

Reply:
Actually, it can - line 114 is outside mutex. It kept failing in a test against RMM to the point that I filed a bug for RMM (by luck it didn't happen with my resource), but then I caught it in the debugger :)